### PR TITLE
BUG: (slow, steady and) correct wins the race: prefer numpy over bottleneck for nanfunctions on float32 arrays

### LINKED ISSUE #11492: Surprising differences in stats due to effects of FITS-endianness
### Description
I was experiencing some really strange issues where a fit would only succeed when executed in a process-pool, but fail in single threaded mode. Digging down, it appears the divergence was due to non-identical output from `sigma_clipped_stats` between single and multi-process execution.
The reason for that boils down to FITS-images being read as dtype `>f4` (big endian, non-native on x86), which when passed directly to bottleneck for mean/median will default to the numpy implementation of `nanmean`/`nanmedian`. In the process of pickling, the array is converted to `<f4` and ends up in the bottleneck implementation.
As far as I can tell, bottleneck performs a straight sum over all elements, while numpy seems to use some form of compensated summation scheme, to decrease numerical error.

### Actual behavior
As bottleneck gets automatically selected by astropy if available, the choice between a accuracy and speed is potentially hidden from the user. This may well be a case of "works as intended" but the chance of reading FITS and unintentionally triggering a conversion to a (on most machines) native array format only under specific circumstances seems rather high. This may cause a lot of headscratching for people that see different results for the same data in seemingly equivalent programs. 

This could plausibly also be considered a bug or unexpected behaviour in bottleneck as the library gives the impression of being a complete drop-in replacement for certain numpy routines.

### Expected behavior
I'm not exactly sure what the best resolution is here. The fact that numpy supports non-native byteorder without explicit conversion/warnings complicates things a bit. Obviously for the user it would be the easiest if all array-operations would have the same results, irrespective of byte-order, but that would possibly cause issues with mmap or performance.

Some kind of warning or note, either when applying `sigma_clipped_stats` to a non-native byteorder or when a FITS file is read, or at least a disclaimer in the docs for `io.fits` about this possible gotcha would help.

### Steps to Reproduce

```python
import numpy as np
import requests
from astropy.stats import sigma_clipped_stats
import warnings

r=requests.get('https://gist.github.com/krachyon/7408a4029fd9cca83f5a8c6ecbeb1173/raw/0ba5cc3e5a0d0eb02a2a4da3d04ec16ee8aa4267/repro_array.npy')
with open('repro_array.npy','wb') as f:
    f.write(r.content)

arr = np.load('repro_array.npy')

with warnings.catch_warnings():
    warnings.filterwarnings('ignore', '.*Input data contains invalid values.*', append=True)

    print('as_is: ', sigma_clipped_stats(arr))
    print('native: ', sigma_clipped_stats(arr.astype(np.float32)))
    print('higher precision: ', sigma_clipped_stats(arr.astype(np.float64)))
    print('')
    print('delta:', np.array(sigma_clipped_stats(arr))-np.array(sigma_clipped_stats(arr.astype(np.float32))))
    print('dtype resolution:', np.finfo(np.float32).resolution)
```

prints

```
as_is:  (2523.1194, 2519.48, 50.65029)
native:  (2523.126220703125, 2519.47998046875, 50.6502799987793)
higher precision:  (2523.1193250563483, 2519.4801025390625, 50.650290905964354)

delta: [-6.83593750e-03  0.00000000e+00  1.14440918e-05]
dtype resolution: 1e-06
```

### System Details
Linux-5.11.11-arch1-1-x86_64-with-glibc2.33
Python 3.9.2 (default, Feb 20 2021, 18:40:11) 
[GCC 10.2.0]
astropy 4.2
bottleneck 1.3.2

### LINKED ISSUE #17185: Unexpected results from sigma_clipped_stats for large np.float32 input arrays
### Description

I'm processing 63 megapixel FITS images using sigma_clipped_stats, and have run into behavior where the endianness of the 32-bit float data is apparently affecting the correctness of the output.

Specifically, data loaded directly out of the fits file `dtype('>f4')` is working correctly. But if I perform numerical operations on the data--or explicitly cast it--such that it becomes `dtype('<f4')` (which is the native float32 on x64: `dtype('<f4')==dtype('float32')`), I get what appear to be incorrect results.

This only happens with large arrays. I haven't determined the exact cutoff, but (4096,4096) appears to work correctly. (6388,9576) does not.

### Expected behavior

I expect `astropy.stats.sigma_clipped_stats(np.array(data,'>f4'))` and `astropy.stats.sigma_clipped_stats(np.array(data,'<f4'))` to produce identical, or at least substantially similar results.

(It appears that there's also a difference in precision calculated in these two cases--while odd, this is not the issue I'm reporting.)

### How to Reproduce

This can be reproduced by generating a random uniform [0-1] array with np.random.random. The mean and median should be approximately 0.5, and the stddev should be approximately 0.288.

The sample code generates arrays of two sizes, and processes them in each endianness. The first three lines have output values that I expect. The last line (larger array, '<f4') gives very different results.

```python
import numpy as np
import astropy, astropy.stats
rand_data = np.random.random(size=(1024,1024))
print(astropy.stats.sigma_clipped_stats(np.array(rand_data,'>f4'), sigma=3, maxiters=5))
print(astropy.stats.sigma_clipped_stats(np.array(rand_data,'<f4'), sigma=3, maxiters=5))
rand_data = np.random.random(size=(6388,9576))
print(astropy.stats.sigma_clipped_stats(np.array(rand_data,'>f4'), sigma=3, maxiters=5))
print(astropy.stats.sigma_clipped_stats(np.array(rand_data,'<f4'), sigma=3, maxiters=5))
```
When  I ran this, I got:
```
(0.4994579, 0.4990998, 0.2887525)
(0.49945005774497986, 0.4990997910499573, 0.28869783878326416)
(0.4999373, 0.49990413, 0.28868327)
(0.2742652893066406, 0.4999041259288788, 0.3462343215942383)
```

### Versions

```
Windows-10-10.0.19045-SP0
Python 3.11.9 (tags/v3.11.9:de54cf5, Apr  2 2024, 10:12:12) [MSC v.1938 64 bit (AMD64)]
astropy 6.1.4
Numpy 1.24.0
pyerfa 2.0.1.4
Scipy 1.11.2
Matplotlib 3.8.0
```

### PR DESCRIPTION
**Pull request by neutrinoceros**
### Description
Fixes #17185
Fixes #11492

maybe we could only fallback on numpy for large arrays, but I don't think we know of a good threshold yet.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
tl;dr -- Does this also fix https://github.com/astropy/astropy/issues/11492 ?

**Review by larrybradley — CHANGES_REQUESTED** [review]
Many thanks, @neutrinoceros!  This looks great.  I think we should mention this behavior somewhere in the Notes section of `SigmaClip` and `sigma_clip`.

**Comment by larrybradley** [comment]
> tl;dr -- Does this also fix https://github.com/astropy/astropy/issues/11492 ?

Yes!  Running the example in that issue with this PR gives deltas of all zeros.

**Comment by neutrinoceros** [comment]
Thanks for pointing it out @larrybradley ! I've added a short note where requested (also in `sigma_clipped_stats`). I also took the opportunity to revamp my test to be less wasteful (in the first iteration, 4 copies were involved, 2 of which were completely unnecessary).

**Review by larrybradley — APPROVED** [review]
Thanks, @neutrinoceros!

**Comment by pllim** [comment]
Let me run the "exotic archs" too, just in case.

**Comment by neutrinoceros** [comment]
switching to draft to prevent merging because I just realized that

```
>>> np.dtype(">f4") == np.dtype("<f4")
False
```
which means that the way I phrased the changelog and docstrings doesn't quite match the actual patch. Since I do not know the exact conditions that hit the bugs in bottlenecks, it seems safer to fix the code and keep the docs as is.

**Comment by neutrinoceros** [comment]
I'll squash the change into a single commit, but for ease of review, here's what's changing:
```patch
diff --git a/astropy/stats/nanfunctions.py b/astropy/stats/nanfunctions.py
index 6901152377..eb483fd23b 100644
--- a/astropy/stats/nanfunctions.py
+++ b/astropy/stats/nanfunctions.py
@@ -98,8 +98,6 @@ if HAS_BOTTLENECK:
         nanstd=np.nanstd,
     )
 
-    _DT_FLOAT32 = np.dtype("float32")
-
     def _dtype_dispatch(func_name):
         # dispatch to bottleneck or numpy depending on the input array dtype
         # this is done to workaround known accuracy bugs in bottleneck
@@ -109,7 +107,7 @@ if HAS_BOTTLENECK:
         # see https://github.com/astropy/astropy/issues/17185
         # see https://github.com/astropy/astropy/issues/11492
         def wrapped(*args, **kwargs):
-            if args[0].dtype == _DT_FLOAT32:
+            if (dt := args[0].dtype).kind == "f" and dt.itemsize <= 4:
                 return np_funcs[func_name](*args, **kwargs)
             else:
                 return bn_funcs[func_name](*args, **kwargs)
```

File: astropy/stats/nanfunctions.py [hunk 1]
```diff
@@ -84,10 +84,40 @@ def _apply_bottleneck(
...
+    def _dtype_dispatch(func_name):
+        # dispatch to bottleneck or numpy depending on the input array dtype
+        # this is done to workaround known accuracy bugs in bottleneck
+        # affecting float32 calculations
+        # see https://github.com/pydata/bottleneck/issues/379
+        # see https://github.com/pydata/bottleneck/issues/462
+        # see https://github.com/astropy/astropy/issues/17185
+        # see https://github.com/astropy/astropy/issues/11492
+        def wrapped(*args, **kwargs):
+            if (dt := args[0].dtype).kind == "f" and dt.itemsize <= 4:
```
Comment (mhvk) [inline astropy/stats/nanfunctions.py:110]: Perhaps overly defensive, but I'd probably only dispatch to bottleneck when the `args[0].dtype.str[1:] == "f8"`...

Comment (neutrinoceros) [reply astropy/stats/nanfunctions.py:110]: Would be fine by me, but I'll let @larrybradley decide what's best here.

Comment (neutrinoceros) [reply astropy/stats/nanfunctions.py]: done. I also updated docstrings and the changelog to reflect this.

**Review by larrybradley — APPROVED** [review]
Thanks again, @neutrinoceros!

**Comment by neutrinoceros** [comment]
Manual backporting is made particularly difficult by previous refactors from #16967
I'm going to dismiss it as not crucial for now, unless @larrybradley disagrees, in which case I'll spend the time needed to backport this.
